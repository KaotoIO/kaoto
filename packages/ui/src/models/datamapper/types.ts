/**
 * The data types mapped from XML Schema +.
 */
export enum Types {
  AnyAtomicType = 'anyAtomicType',
  AnyType = 'anyType',
  AnyURI = 'anyURI',
  String = 'string',
  Float = 'float',
  Double = 'double',
  Integer = 'integer',
  Decimal = 'decimal',
  Boolean = 'boolean',
  Date = 'date',
  DateTime = 'dateTime',
  Time = 'time',
  QName = 'QName',
  Duration = 'duration',
  DayTimeDuration = 'dayTimeDuration',
  NCName = 'NCName',
  // not XSD predefined but appear in XPath function signature
  Numeric = 'numeric',
  Item = 'item()',
  Element = 'element()',
  Node = 'node()',
  DocumentNode = 'document-node()',
  // custom types
  Container = 'Container',
  Array = 'Array',
}
