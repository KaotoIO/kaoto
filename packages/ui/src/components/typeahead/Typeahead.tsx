import {
  Button,
  MenuToggle,
  MenuToggleElement,
  Select,
  SelectList,
  SelectOption,
  TextInputGroup,
  TextInputGroupMain,
  TextInputGroupUtilities,
} from '@patternfly/react-core';
import { TimesIcon } from '@patternfly/react-icons';
import {
  FormEvent,
  FunctionComponent,
  MouseEventHandler,
  Ref,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react';
import { isDefined } from '../../utils';
import { TypeaheadProps } from './Typeahead.types';

export const CREATE_NEW_ITEM = 'create-new-with-name';

const DEFAULT_POPPER_PROPS = {
  position: 'end',
  preventOverflow: true,
} as const;

export const Typeahead: FunctionComponent<TypeaheadProps> = ({
  selectedItem,
  items: itemsProps,
  id,
  placeholder = 'Select or write an option',
  onChange,
  onCleanInput,
  'aria-label': ariaLabel,
  'data-testid': dataTestId,
  onCreate,
  onCreatePrefix,
  disabled = false,
}) => {
  const [filter, setFilter] = useState<string>(selectedItem?.name ?? '');
  const [inputValue, setInputValue] = useState<string>(selectedItem?.name ?? '');
  const [isOpen, setIsOpen] = useState(false);
  const inputRef = useRef<HTMLInputElement>(null);
  const selectedOptionRef = useRef<HTMLSelectElement>(null);

  const items = useMemo(() => {
    const isValueInArray = isDefined(itemsProps.find((item) => item.name === selectedItem?.name));
    const localArray = itemsProps.slice();
    if (isValueInArray) {
      return localArray;
    }
    if (selectedItem?.name && selectedItem?.value) {
      localArray.unshift({ name: selectedItem.name, value: selectedItem.value });
    }
    return localArray;
  }, [itemsProps, selectedItem?.name, selectedItem?.value]);

  useEffect(() => {
    if (selectedItem?.name) {
      setInputValue(selectedItem.name);
    }
  }, [selectedItem]);

  const onItemChanged = useCallback(
    (_event: unknown, name: string | number | undefined) => {
      if (name === CREATE_NEW_ITEM) {
        onCreate?.(name, inputValue);
        setIsOpen(false);
        return;
      } else if (!isDefined(name)) {
        onChange?.(undefined);
        setInputValue('');
        setIsOpen(false);
        return;
      }

      const localItem = items.find((item) => item.value === name);
      setInputValue(localItem?.name ?? '');
      setIsOpen(false);

      if (name !== selectedItem?.name) {
        onChange?.(localItem);
      }
    },
    [onChange, items, selectedItem?.name, onCreate, inputValue],
  );

  const onToggleClick: MouseEventHandler<HTMLDivElement | HTMLButtonElement> = async (event) => {
    event.stopPropagation();
    if (isOpen) {
      setIsOpen(false);
      return;
    }

    setFilter('');
    setIsOpen(true);

    requestAnimationFrame(() => {
      selectedOptionRef.current?.scrollIntoView({ behavior: 'smooth', block: 'nearest', inline: 'nearest' });
      inputRef.current?.focus();
    });
  };

  const onInputClick: MouseEventHandler<HTMLDivElement> = (event) => {
    event.stopPropagation();
    setFilter(inputValue);
    setIsOpen(true);
  };

  const onTextInputChange = (_event: FormEvent<HTMLInputElement>, value: string) => {
    setFilter(value);
    setInputValue(value);
  };

  const onTextInputClear = () => {
    setFilter('');
    setInputValue('');
    onCleanInput?.();
    inputRef.current?.focus();
  };

  const filteredItems = useMemo(() => {
    const lowerFilterValue = filter.toLowerCase();
    return items.filter((item) => {
      if (!lowerFilterValue) {
        return true;
      }

      const hasNameMatch = item.name?.toLowerCase().includes(lowerFilterValue);
      const hasDescriptionMatch = item.description?.toLowerCase().includes(lowerFilterValue);

      return hasNameMatch || hasDescriptionMatch;
    });
  }, [filter, items]);

  const toggle = (toggleRef: Ref<MenuToggleElement>) => (
    <MenuToggle
      isFullWidth
      ref={toggleRef}
      onClick={onToggleClick}
      isExpanded={isOpen}
      isDisabled={disabled}
      variant="typeahead"
      aria-label={`${ariaLabel} toggle`}
      id={id}
    >
      <TextInputGroup isPlain>
        <TextInputGroupMain
          autoComplete="off"
          id={`${id}-typeahead-select-input`}
          aria-label={ariaLabel}
          data-testid={`${dataTestId}-typeahead-select-input`}
          ref={inputRef}
          placeholder={placeholder}
          onClick={onInputClick}
          value={inputValue}
          onChange={onTextInputChange}
        />

        <TextInputGroupUtilities>
          {!!inputValue && (
            <Button
              variant="plain"
              onClick={onTextInputClear}
              aria-label="Clear input value"
              data-testid={`${dataTestId}__clear`}
              icon={<TimesIcon aria-hidden />}
            />
          )}
        </TextInputGroupUtilities>
      </TextInputGroup>
    </MenuToggle>
  );

  return (
    <Select
      id={`${id}-typeahead-select`}
      data-testid={`${dataTestId}-typeahead-select`}
      isScrollable
      shouldFocusToggleOnSelect
      isOpen={isOpen}
      selected={selectedItem?.name}
      onSelect={onItemChanged}
      toggle={toggle}
      onOpenChange={(isOpen) => setIsOpen(isOpen)}
      popperProps={DEFAULT_POPPER_PROPS}
    >
      <SelectList>
        {filteredItems.map((item) => (
          <SelectOption
            key={item.name}
            value={item.value}
            description={item.description}
            aria-label={`option ${item.name.toLocaleLowerCase()}`}
            isSelected={item.name === selectedItem?.name}
            ref={item.name === selectedItem?.name ? selectedOptionRef : undefined}
          >
            {item.name}
          </SelectOption>
        ))}

        {filteredItems.length === 0 && <SelectOption isDisabled>No items found</SelectOption>}

        {onCreate && (
          <SelectOption value={CREATE_NEW_ITEM} aria-label={`option ${CREATE_NEW_ITEM.toLocaleLowerCase()}`}>
            Create new {onCreatePrefix} {inputValue ? `'${inputValue}'` : ''}
          </SelectOption>
        )}
      </SelectList>
    </Select>
  );
};
