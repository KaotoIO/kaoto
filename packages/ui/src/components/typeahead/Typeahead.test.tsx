import { render, screen, fireEvent, act } from '@testing-library/react';
import { CREATE_NEW_ITEM, Typeahead } from './Typeahead';
import { TypeaheadProps } from './Typeahead.types';

const mockItems = [
  { name: 'Item 1', value: 'item1', description: 'Description 1' },
  { name: 'Item 2', value: 'item2', description: 'Description 2' },
];

const defaultProps: TypeaheadProps = {
  selectedItem: undefined,
  items: mockItems,
  id: 'test-typeahead',
  onChange: jest.fn(),
  onCleanInput: jest.fn(),
  'aria-label': 'Typeahead',
  'data-testid': 'typeahead',
};

describe('Typeahead', () => {
  beforeEach(() => {
    jest.useFakeTimers();
  });

  afterEach(() => {
    act(() => {
      jest.runAllTimers();
    });
    jest.useRealTimers();
  });

  it('should renders the Typeahead component', async () => {
    const { container } = render(<Typeahead {...defaultProps} />);

    expect(container).toMatchSnapshot();
  });

  it('should renders the Typeahead component with disabled button', async () => {
    const { container } = render(<Typeahead {...defaultProps} disabled={true} />);

    expect(container).toMatchSnapshot();
  });

  it('should opens the dropdown when the toggle is clicked', async () => {
    render(<Typeahead {...defaultProps} />);
    const toggle = screen.getByLabelText('Typeahead toggle');

    await act(async () => {
      fireEvent.click(toggle);
    });

    expect(screen.getByTestId('typeahead-typeahead-select')).toBeInTheDocument();
  });

  it('should filters items based on input value', async () => {
    render(<Typeahead {...defaultProps} />);

    const input = screen.getByPlaceholderText('Select or write an option');
    await act(async () => {
      fireEvent.click(input);
    });

    await act(async () => {
      fireEvent.change(input, { target: { value: 'Item 1' } });
    });

    expect(screen.getByText('Item 1')).toBeInTheDocument();
    expect(screen.queryByText('Item 2')).not.toBeInTheDocument();
  });

  it('should display no items found when no items match the input value', async () => {
    render(<Typeahead {...defaultProps} />);

    const input = screen.getByPlaceholderText('Select or write an option');
    await act(async () => {
      fireEvent.click(input);
    });

    await act(async () => {
      fireEvent.change(input, { target: { value: 'Non-existent Item' } });
    });

    expect(screen.getByText('No items found')).toBeInTheDocument();
  });

  it('should calls onChange when an item is selected', async () => {
    render(<Typeahead {...defaultProps} />);
    const toggle = screen.getByLabelText('Typeahead toggle');

    await act(async () => {
      fireEvent.click(toggle);
    });

    const option = screen.getByText('Item 1');
    act(() => {
      fireEvent.click(option);
    });

    expect(defaultProps.onChange).toHaveBeenCalledWith(mockItems[0]);
  });

  it('should calls onCleanInput when clear button is clicked', async () => {
    render(<Typeahead {...defaultProps} />);
    const input = screen.getByPlaceholderText('Select or write an option');

    await act(async () => {
      fireEvent.change(input, { target: { value: 'Item 1' } });
    });

    const clearButton = screen.getByLabelText('Clear input value');
    fireEvent.click(clearButton);

    expect(defaultProps.onCleanInput).toHaveBeenCalled();
  });

  it('should allow users to create a new item if the onCreate callback is set', async () => {
    render(<Typeahead {...defaultProps} onCreate={jest.fn()} onCreatePrefix="multiverse" />);

    const input = screen.getByPlaceholderText('Select or write an option');
    await act(async () => {
      fireEvent.click(input);
    });

    expect(screen.getByText('Create new multiverse')).toBeInTheDocument();
  });

  it('should allow users to create a new item if the onCreate callback is set and there is a value', async () => {
    render(<Typeahead {...defaultProps} onCreate={jest.fn()} onCreatePrefix="brick" />);

    const input = screen.getByPlaceholderText('Select or write an option');
    await act(async () => {
      fireEvent.click(input);
    });

    await act(async () => {
      fireEvent.change(input, { target: { value: 'in the wall' } });
    });

    const createNewElement = screen.getByLabelText(`option ${CREATE_NEW_ITEM.toLocaleLowerCase()}`);

    expect(createNewElement).toBeInTheDocument();
    expect(createNewElement).toHaveTextContent("Create new brick 'in the wall'");
  });
});
