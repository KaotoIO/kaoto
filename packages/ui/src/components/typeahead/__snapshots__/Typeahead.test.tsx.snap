// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typeahead should renders the Typeahead component 1`] = `
<div>
  <div
    class="pf-v6-c-menu-toggle pf-m-full-width pf-m-typeahead"
    id="test-typeahead"
  >
    <div
      class="pf-v6-c-text-input-group pf-m-plain"
    >
      <div
        autocomplete="off"
        class="pf-v6-c-text-input-group__main"
        data-testid="typeahead-typeahead-select-input"
        id="test-typeahead-typeahead-select-input"
      >
        <span
          class="pf-v6-c-text-input-group__text"
        >
          <input
            aria-label="Typeahead"
            class="pf-v6-c-text-input-group__text-input"
            placeholder="Select or write an option"
            type="text"
            value=""
          />
        </span>
      </div>
      <div
        class="pf-v6-c-text-input-group__utilities"
      />
    </div>
    <button
      aria-expanded="false"
      aria-label="Typeahead toggle"
      class="pf-v6-c-menu-toggle__button"
      data-ouia-component-id="OUIA-Generated-MenuToggle-typeahead-1"
      data-ouia-component-type="PF6/MenuToggle"
      data-ouia-safe="true"
      tabindex="-1"
      type="button"
    >
      <span
        class="pf-v6-c-menu-toggle__controls"
      >
        <span
          class="pf-v6-c-menu-toggle__toggle-icon"
        >
          <svg
            aria-hidden="true"
            class="pf-v6-svg"
            fill="currentColor"
            height="1em"
            role="img"
            viewBox="0 0 320 512"
            width="1em"
          >
            <path
              d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
            />
          </svg>
        </span>
      </span>
    </button>
  </div>
</div>
`;

exports[`Typeahead should renders the Typeahead component with disabled button 1`] = `
<div>
  <div
    class="pf-v6-c-menu-toggle pf-m-full-width pf-m-disabled pf-m-typeahead"
    disabled=""
    id="test-typeahead"
  >
    <div
      class="pf-v6-c-text-input-group pf-m-plain"
    >
      <div
        autocomplete="off"
        class="pf-v6-c-text-input-group__main"
        data-testid="typeahead-typeahead-select-input"
        id="test-typeahead-typeahead-select-input"
      >
        <span
          class="pf-v6-c-text-input-group__text"
        >
          <input
            aria-label="Typeahead"
            class="pf-v6-c-text-input-group__text-input"
            placeholder="Select or write an option"
            type="text"
            value=""
          />
        </span>
      </div>
      <div
        class="pf-v6-c-text-input-group__utilities"
      />
    </div>
    <button
      aria-expanded="false"
      aria-label="Typeahead toggle"
      class="pf-v6-c-menu-toggle__button"
      data-ouia-component-id="OUIA-Generated-MenuToggle-typeahead-2"
      data-ouia-component-type="PF6/MenuToggle"
      data-ouia-safe="true"
      tabindex="-1"
      type="button"
    >
      <span
        class="pf-v6-c-menu-toggle__controls"
      >
        <span
          class="pf-v6-c-menu-toggle__toggle-icon"
        >
          <svg
            aria-hidden="true"
            class="pf-v6-svg"
            fill="currentColor"
            height="1em"
            role="img"
            viewBox="0 0 320 512"
            width="1em"
          >
            <path
              d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
            />
          </svg>
        </span>
      </span>
    </button>
  </div>
</div>
`;
